### PR TITLE
MediaPlayer Workaround code

### PR DESCRIPTION
**Pull request by junmin-kim**
This codes should be refactoried

### CODE REVIEW

File: apps/examples/cxxtest/MediaPlayer.cxx [hunk 1]
```diff
@@ -87,6 +89,21 @@ namespace Media
 		enqueue([this]() {_sync(); });
 		cvSync.wait(lock);
 
+		pcmOut = pcm_open(0, 0, PCM_OUT, NULL);
+		if (pcm_is_ready(pcmOut) == 0) {
+			std::cout << "MediaPlayer::preare : pcm is not ready" << std::endl;
+			return PLAYER_ERROR;
+		}
+
+		fd = open("/mnt/record", O_RDONLY);
```
Comment (jsdosa) [inline apps/examples/cxxtest/MediaPlayer.cxx]: use c++ apis. 
example fstream.

File: apps/examples/cxxtest/MediaPlayer.hpp [hunk 2]
```diff
@@ -73,7 +73,12 @@ namespace Media
 		void _pause();
 		void _stop();
 
-	private:		
+	private:
+		struct pcm *pcmOut;
+		int fd;
```
Comment (jsdosa) [inline apps/examples/cxxtest/MediaPlayer.hpp]: use c++ apis. 
fstream.

File: apps/examples/cxxtest/MediaPlayer.cxx [hunk 3]
```diff
@@ -87,6 +89,21 @@ namespace Media
 		enqueue([this]() {_sync(); });
 		cvSync.wait(lock);
 
+		pcmOut = pcm_open(0, 0, PCM_OUT, NULL);
+		if (pcm_is_ready(pcmOut) == 0) {
```
Comment (Taejun-Kwon) [inline apps/examples/cxxtest/MediaPlayer.cxx]: Below is better.
if (!pcm_is_ready(pcmOut)) {
    LOG ("Player : prepare error .... ...... %d\n", pcm_get_file_descrtiptor());
}

File: apps/examples/cxxtest/MediaPlayer.cxx [hunk 4]
```diff
@@ -87,6 +89,21 @@ namespace Media
...
+
+		fd = open("/mnt/record", O_RDONLY);
+		if (fd < 0) {
+			std::cout << "MediaPlayer::prepare : open fail" << std::endl;
+			return PLAYER_ERROR;
+		}
+
+		size = pcm_frames_to_bytes(pcmOut, pcm_get_buffer_size(pcmOut));
+		buffer = (char *)malloc(size);
+
```
Comment (Taejun-Kwon) [inline apps/examples/cxxtest/MediaPlayer.cxx]: Null check needed.

File: apps/examples/cxxtest/MediaPlayer.cxx [hunk 5]
```diff
@@ -218,12 +240,23 @@ namespace Media
 			unique_lock<mutex> lock(*qMtx);
 
 			if (cmdQueue.empty()) {
-				cvQueue.wait(lock);
+				if (curState == PLAYER_STATE_PLAYING) {
+					int num_read = read(fd, buffer, size);
+					if (num_read > 0) {
+						pcm_writei(pcmOut, buffer, pcm_bytes_to_frames(pcmOut, num_read));
```
Comment (Taejun-Kwon) [inline apps/examples/cxxtest/MediaPlayer.cxx]: ret = pcm_writei();
if (ret == -EPIPE) {
    ret = pcm_prepare();
    if (ret != OK) {
          LOG();
          return SOME_ERRORS;
    }
    ret = pcm_writei(); //try once more
}

File: apps/examples/cxxtest/MediaPlayer.cxx [hunk 6]
```diff
@@ -218,12 +240,23 @@ namespace Media
...
+					} else {
+						curState = PLAYER_STATE_STOP;
```
Comment (Taejun-Kwon) [inline apps/examples/cxxtest/MediaPlayer.cxx]: this.stop();
user_cb(); //notify that playback completed.

File: apps/examples/cxxtest/MediaPlayer.cxx [hunk 7]
```diff
@@ -218,12 +240,23 @@ namespace Media
...
+					}
+					continue;
```
Comment (Taejun-Kwon) [inline apps/examples/cxxtest/MediaPlayer.cxx]: why continue needed?

File: apps/examples/cxxtest/MediaPlayer.cxx (hunk 6 above)
Comment (Taejun-Kwon) [reply apps/examples/cxxtest/MediaPlayer.cxx]: I think 2 cb needed, one is for 'state change', one is for 'playback completed'

**Comment by Taejun-Kwon** [comment]
Let's think about simple music player. This will be helpful to design.
int main() {
      position = 0;
      trackList[i] = getTrackList("/mnt/mp3");
      playCurrent();
}
void playCurrent() {
      DataSource data = new DataSource(2, 16000, 16, trackList(position));
      MediaPlayer mp = new MediaPlayer(data.....); 
      ret = mp.prepare();
      if (ret == PLAYER_OK) {
           mp.setVolume(0); // prevent noise
           mp.start();
           if (is_playing()) {
                 mp.setVolume(1.0f);
           }
      }
}
void onCompletion() {
    data.delete();
    mp.delete();
    position++;
    playCurrent(position);
}
void onMediaStateChanged(int state) {
    switch(state)  {
       case IDLE : 
                turn_off_light();
                break;
      case PLAYING :
               turn_on_light();
               isPlaying = true;
               break;
      default :
              ERROR!!!
}

void onButtonPressed() {
     if (left) {
         mp.pause();
     } else if(right) {
          mp.resume();
     } else {
          mp.stop();
     }
}

**Comment by junmin-kim** [comment]
@Taejun-Kwon @jsdosa 
Should we support multiple user cb func?

File: apps/examples/cxxtest/MediaPlayer.hpp (hunk 2 above)
Comment (junmin-kim) [reply apps/examples/cxxtest/MediaPlayer.hpp]: Done

File: apps/examples/cxxtest/MediaPlayer.cxx [hunk 8]
```diff
@@ -218,12 +243,33 @@ namespace Media
 			unique_lock<mutex> lock(*qMtx);
 
 			if (cmdQueue.empty()) {
-				cvQueue.wait(lock);
+				if (curState == PLAYER_STATE_PLAYING) {
+					ifs.read(buffer, size);
```
Comment (Taejun-Kwon) [inline apps/examples/cxxtest/MediaPlayer.cxx]: let's align recorder & player
for example, fs_in for recorder / fs_out for player...

Comment (jsdosa) [reply apps/examples/cxxtest/MediaPlayer.cxx]: ok, i will change,
but that is file operation. recorder side is fs_out correct.

**Comment by junmin-kim** [comment]
Please merge this PR. After merge, I'll change what we discussed

**Pull request merged**